### ISSUE TITLE
Close connections to the hg server once the process pool is terminated

### ISSUE THREAD

**Issue by marco-c** [mozilla/bugbug#769]
Similar to ec6ffeb38f6778f95f5a76ebc2ba331d27cf5cac, but for the process pool.

**Comment by harsh020** [comment]
Hey, as there are no PRs open for this issue can I take it? If yes, I need a bit help as this will be my first open source contribution. (I'm reading the contribution docs). Just a question, the clear function (@ line 757), tends to open the hg server, we need to close it there?
The clone function (@ line 776), is it building the repo if not already existing, so we also need to close the hg server there also?

Thanks.

**Comment by harsh020** [comment]
@marco-c Hey, as there is no PR for this issue I would like to take it.
As this is my very first open source contribution I would be very grateful if you helped me out. I went through the file repository.py but still am not clear where should I close the hg server, is it in the clear function at line 757?

**Comment by marco-c** [comment]
Yes, it's 2.
In 1 we are already closing the hg server, as we are using a context manager https://github.com/mozilla/bugbug/blob/17ff6ed31569fa893e13b8a1140bd1f458765a70/bugbug/repository.py#L758.

@harsh020 note that this is a good-first-issue as it is self-contained, but it is not a good-first-issue if you don't have experience with Python and knowledge of how multiprocessing works. Feel free to propose a PR if you manage to fix it, or to switch to some other good-first-issue if this one is too hard.

> I think it would be the second one, if so, should I use the global variable and append it every time , like in the link you provided and close all the servers before returning the commits?

If you use a global variable, each process will have its own copy, so this won't work.

**Comment by harsh020** [comment]
> In 1 we are already closing the hg server, as we are using a context manager
@marco-c Sorry, I did not notice that.

> if you don't have experience with Python and knowledge of how multiprocessing works. 
Well, I've been working with python for about two years (yup, not too much experience) and I'll read more about multiprocessing, not a problem.

> If you use a global variable, each process will have its own copy, so this won't work.
I'm not thinking to declare a global variable, but the `_init` function itself has a `global HG`, should I simply append them in a list and close all of the servers before returning the commits?

**Comment by marco-c** [comment]
> yup, not much experience, but I wanna learn, and I'll read more about multiprocessing, not a problem.

OK! Note: for `multiprocessing` I'm not just talking about the `multiprocessing` module, but concepts about working with multiple processes in general, if you don't know them you'll need to read about them too.

> I'm not thinking to declare a global variable, but the _init function itself has a global HG, should I simply append them in a list and close all of the servers before returning the commits?

Sure, you can do that, but you need to share the list between all processes.

**Comment by harsh020** [comment]
> for multiprocessing I'm not just talking about the multiprocessing module, but concepts about working with multiple processes in general

Yes, I'm on it. Threading (drawbacks bcs of GIL) and multiprocessing.

> Sure, you can do that, but you need to share the list between all processes.

I was thinking of creating a global list to make it accessible to all process like in the link you provided.

**Comment by marco-c** [comment]
> I was thinking of creating a global list to make it accessible to all process like in the link you provided.

As I said earlier, a global list won't work because global variables are not shared between processes. Each process will have its own copy.

**Comment by harsh020** [comment]
@marco-c Won't just declaring the list outside the function do the job? Thats interesting.

Thanks.

**Comment by marco-c** [comment]
@harsh020 I told you this requires a bit of experience with multiprocessing concepts :)

**Comment by ayush-1506** [comment]
@harsh020 Multiprocessing has `Value` and `Array` that gets shared between processes. Global variables aren't shared since each process has it's own copy.

**Comment by harsh020** [comment]
@marco-c Sorry for the delayed reply, I'm still studying the concepts and won't open any PR without informing you the approach. 

> I told you this requires a bit of experience with multiprocessing concepts :)

Yup, very true.

Thanks.

**Comment by harsh020** [comment]
> Multiprocessing has Value and Array that gets shared between processes. Global variables aren't shared since each process has it's own copy/address space.

@ayush-1506 Yup, I read about it, thats an amazing (yet annoying ) thing.
Thanks for telling me.
